lib/puppet/defaults.rb

```diff
@@ -676,37 +676,11 @@ def self.initialize_default_settings!(settings)
     :environment_timeout => {
       :default    => "0",
       :type       => :ttl,
-      :desc       => "How long the Puppet master should cache data it loads from an
-      environment.
-
-      A value of `0` will disable caching. This setting can also be set to
-      `unlimited`, which will cache environments until the master is restarted
-      or told to refresh the cache.
-
-      You should change this setting once your Puppet deployment is doing
-      non-trivial work. We chose the default value of `0` because it lets new
-      users update their code without any extra steps, but it lowers the
-      performance of your Puppet master.
-
-      We recommend setting this to `unlimited` and explicitly refreshing your
-      Puppet master as part of your code deployment process.
-
-      * With Puppet Server, you should refresh environments by calling the
-        `environment-cache` API endpoint. See the docs for the Puppet Server
-        [administrative API](https://puppet.com/docs/puppetserver/latest/admin-api/v1/environment-cache.html).
-
-      Any value other than `0` or `unlimited` is deprecated, since most Puppet
-      servers use a pool of Ruby interpreters which all have their own cache
-      timers. When these timers drift out of sync, agents can be served
-      inconsistent catalogs.",
-      :hook => proc do |val|
-        unless [0, 'unlimited', Float::INFINITY].include?(val)
-          Puppet.deprecation_warning(<<-WARNING)
-Fine grained control of environment timeouts is deprecated,
-please use `environment_ttl` instead.
-          WARNING
-        end
-      end
+      :desc       => "This setting is not supported, use `environment_ttl` instead.",
+      :hook => proc { |value|
+        Puppet.warning("Setting 'environment_timeout' is no longer supported. Use 'environment_ttl' instead.")
+        Puppet[:environment_ttl] = value unless Puppet.settings.set_by_config?(:environment_ttl)
```

Comment:
@npwalker This implements what we talked about today. If `environment_timeout` is set in `puppet.conf` then we issue a warning (that can't be disabled) and set `environment_ttl` instead. It's a little strange saying the setting is not supported and then using its value... I thought about changing the warning to a hard error, e.g `raise ArgumentError("Setting `environment_timeout` is no longer supported. Use `environment_ttl` instead.)`. Also this PR will silently ignore `environment_timeout` in `environment.conf`. I don't know how common that is? Thoughts? /cc @Magisus @justinstoller ?

Reply:
Yeah, setting the ttl to the timeout is a little weird, is that a thing that you and Nick decided on?

I I'm just wondering if there's a range of common values folks use that are valid for both a ttl and a timeout - I would imagine most timeouts when applied as ttls would cause things to be effectively cached forever. I don't know if erring on the side of caching things for longer than users intended is better than not caching when users wanted some caching. Or whether doing something with the value will hide that the system no longer works in a way that matches how users understand it.

Reply:
I know we don't encourage folks to set it per environment in our [bootstrap control-repo](https://github.com/puppetlabs/control-repo/blob/production/environment.conf), and that InfraCore doesn't use it. I _think_ that's indicative of how it's (not) used more widely.

Reply:
Yea the warning should inform the user what is happening.  Something like `'environment_timeout' is now an alias to 'environment_ttl'.  Avoid this warning by setting 'environment_ttl' instead.`

I'm not sure we should silently ignore setting environment_timeout in environment.conf.  The workaround provided to customers that wanted to reduce memory usage was to go through git machinations to set the environment_timeout to 0 in  non-production environments.   If we ignore it there I suspect some users will get completely different caching behavior ( either all on or all off ) and get severely degraded perf without a clue as to why.  

I don't think it's about encouraging/discouraging use of environment_timeout in environment.conf.  It's just that trying to set it there gives you conflicts when you merge long standing branches / feature branches into production that most people don't want to reason about.  

as to Justin's point "I don't know if erring on the side of caching things for longer than users intended is better than not caching when users wanted some caching."  I don't think we'll be changing the memory profile in an unexpected way.  If you set the environment_timeout to 3 minutes and then you push up your dev branch, compile a catalog against it and then  compile a catalog against it again 2 minutes 30 seconds later and 5 minutes later.  Do you really want the cache to be evicted between those two compilations?  It doesn't change the memory usage significantly other reducing GC on the unnecessarily evicted cache and actually evicting the cache at the end sooner under ttl than timeout.

Reply:
The PR has been changed to avoid the aliasing issue.